x-pack/platform/plugins/shared/onechat/server/routes/tools.ts

```diff
@@ -6,35 +6,162 @@
  */
 
 import { schema } from '@kbn/config-schema';
+import { ToolType } from '@kbn/onechat-common';
 import type { RouteDependencies } from './types';
 import { getHandlerWrapper } from './wrap_handler';
 import { toolToDescriptor } from '../services/tools/utils/tool_conversion';
-import type { ListToolsResponse } from '../../common/http_api/tools';
+import type {
+  ListToolsResponse,
+  GetToolResponse,
+  DeleteToolResponse,
+  CreateToolPayload,
+  UpdateToolPayload,
+  CreateToolResponse,
+  UpdateToolResponse,
+} from '../../common/http_api/tools';
 import { apiPrivileges } from '../../common/features';
+import {
+  configurationSchema as esqlConfigSchema,
+  configurationUpdateSchema as esqlConfigUpdateSchema,
+} from '../services/tools/esql/schemas';
 
 export function registerToolsRoutes({ router, getInternalServices, logger }: RouteDependencies) {
   const wrapHandler = getHandlerWrapper({ logger });
 
-  router.post(
+  // list tools API
+  router.get(
     {
-      path: '/internal/onechat/tools',
+      path: '/api/chat/tools',
       security: {
         authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
       },
       validate: false,
     },
     wrapHandler(async (ctx, request, response) => {
       const { tools: toolService } = getInternalServices();
-      const registry = toolService.registry.asScopedPublicRegistry({ request });
+      const registry = await toolService.getRegistry({ request });
       const tools = await registry.list({});
       return response.ok<ListToolsResponse>({
         body: {
-          tools: tools.map(toolToDescriptor),
+          results: tools.map(toolToDescriptor),
         },
       });
     })
   );
 
+  // get tool by ID
+  router.get(
+    {
+      path: '/api/chat/tools/{id}',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
+      },
+      validate: {
+        params: schema.object({
+          id: schema.string(),
+        }),
+      },
+    },
+    wrapHandler(async (ctx, request, response) => {
+      const { id } = request.params;
+      const { tools: toolService } = getInternalServices();
+      const registry = await toolService.getRegistry({ request });
+      const tool = await registry.get(id);
+      return response.ok<GetToolResponse>({
+        body: toolToDescriptor(tool),
+      });
+    })
+  );
+
+  // create tool
+  router.post(
+    {
+      path: '/api/chat/tools',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.manageOnechat] },
+      },
+      validate: {
+        body: schema.object({
+          id: schema.string(),
+          type: schema.maybe(
+            schema.oneOf([schema.literal(ToolType.esql), schema.literal(ToolType.builtin)])
```

Comment:
is that right we want to allow builtin tools to be user created?

Reply:
We can't, it's just that the validation is handled lower, in the service layer. TBH I did that because we're more in control of error messages in the tool registry that we are with config-schema, and to as much as possible in a single  place.